tests/components/climate/test_melissa.py

```diff
@@ -15,244 +14,393 @@
 )
-from tests.common import get_test_home_assistant, load_fixture
-
-
-class TestMelissa(unittest.TestCase):
-    """Tests for Melissa climate."""
-
-    def setUp(self):  # pylint: disable=invalid-name
-        """Set up test variables."""
-        self.hass = get_test_home_assistant()
-        self._serial = '12345678'
-
-        self.api = Mock()
-        self.api.fetch_devices.return_value = json.loads(load_fixture(
-            'melissa_fetch_devices.json'
-        ))
-        self.api.cur_settings.return_value = json.loads(load_fixture(
-            'melissa_cur_settings.json'
-        ))
-        self.api.status.return_value = json.loads(load_fixture(
-            'melissa_status.json'
-        ))
-        self.api.STATE_OFF = 0
-        self.api.STATE_ON = 1
-        self.api.STATE_IDLE = 2
-
-        self.api.MODE_AUTO = 0
-        self.api.MODE_FAN = 1
-        self.api.MODE_HEAT = 2
-        self.api.MODE_COOL = 3
-        self.api.MODE_DRY = 4
-
-        self.api.FAN_AUTO = 0
-        self.api.FAN_LOW = 1
-        self.api.FAN_MEDIUM = 2
-        self.api.FAN_HIGH = 3
-
-        self.api.STATE = 'state'
-        self.api.MODE = 'mode'
-        self.api.FAN = 'fan'
-        self.api.TEMP = 'temp'
-
-        device = self.api.fetch_devices()[self._serial]
-        self.thermostat = melissa.MelissaClimate(
-            self.api, device['serial_number'], device)
-        self.thermostat.update()
-
-    def tearDown(self):  # pylint: disable=invalid-name
-        """Teardown this test class. Stop hass."""
-        self.hass.stop()
-
-    @patch("homeassistant.components.climate.melissa.MelissaClimate")
-    def test_setup_platform(self, mocked_thermostat):
-        """Test setup_platform."""
-        device = self.api.fetch_devices()[self._serial]
-        thermostat = mocked_thermostat(self.api, device['serial_number'],
+from tests.common import load_fixture, mock_coro_func
+
+_SERIAL = "12345678"
+
+
+def melissa_mock():
+    """Mocked melissa api."""
+
+    api = Mock()
+    api.async_fetch_devices = mock_coro_func(
+        return_value=json.loads(load_fixture('melissa_fetch_devices.json')))
+    api.async_status = mock_coro_func(return_value=json.loads(load_fixture(
+        'melissa_status.json')))
+    api.async_cur_settings = mock_coro_func(
+        return_value=json.loads(load_fixture('melissa_cur_settings.json')))
+
+    api.async_send = mock_coro_func(return_value=True)
+
+    api.STATE_OFF = 0
+    api.STATE_ON = 1
+    api.STATE_IDLE = 2
+
+    api.MODE_AUTO = 0
+    api.MODE_FAN = 1
+    api.MODE_HEAT = 2
+    api.MODE_COOL = 3
+    api.MODE_DRY = 4
+
+    api.FAN_AUTO = 0
+    api.FAN_LOW = 1
+    api.FAN_MEDIUM = 2
+    api.FAN_HIGH = 3
+
+    api.STATE = 'state'
+    api.MODE = 'mode'
+    api.FAN = 'fan'
+    api.TEMP = 'temp'
+    return api
+
+
+async def test_setup_platform(hass):
+    """Test setup_platform."""
+    with patch("homeassistant.components.climate.melissa.MelissaClimate"
```

Comment:
It would be more robust and require less tests if we would write the tests more like integration tests as is standard in home assistant. Ie set up the component and platform with config, let the core add the entity and then assert that the state machine has the correct state, all while patching the library api (I/O) as needed. Also test the actuator methods.

Most existing tests should be rewritten, but some can probably just be removed as that functionality will be covered by the other tests.

Reply:
It's ok to wait with the test migration. We can merge when tests pass.

Reply:
Ok, I'll add the test-migration as soon as possible, I agree 100% that the tests in this component is not correctly implemented